### ISSUE TITLE
119 chrome makes PDFs huge and slow

### ISSUE THREAD

**Issue by piskov** [hardkoded/puppeteer-sharp#2381]
If chrome 119.0.6045.105 is used, we noticed more than 2× the time and 3× the size (508 KB vs 1855 KB) for generated PDFs  compared to any previous version like 118, 117 or 115

May be #2354 should be reverted

PuppeteerSharp 13.0.2
.net 6

**Comment by kblok** [comment]
@piskov I'm getting the same size with this piece of code. Could you help me reproduce that?

```
using System;
using System.Linq;
using System.IO;
using System.Threading.Tasks;
using PuppeteerSharp;

namespace PuppeteerSharpPdfDemo
{
    class MainClass
    {
        public static async Task Main(string[] args)
        {
            Console.WriteLine("Downloading browsers");

            using var browserFetcher = new BrowserFetcher(SupportedBrowser.Chrome);
            var chrome118 = await browserFetcher.DownloadAsync("118.0.5993.70");
            var chrome119 = await browserFetcher.DownloadAsync("119.0.5997.0");

            Console.WriteLine("Navigating");
            await using (var browser = await Puppeteer.LaunchAsync(new()
                         {
                             ExecutablePath = chrome118.GetExecutablePath(),
                         }))
            {
                await using var page = await browser.NewPageAsync();
                await page.GoToAsync("https://www.whatismybrowser.com/");

                Console.WriteLine("Generating PDF");
                await page.PdfAsync(Path.Combine(Directory.GetCurrentDirectory(), "118.pdf"));

                Console.WriteLine("Export completed");
            }

            await using (var browser = await Puppeteer.LaunchAsync(new()
                         {
                             ExecutablePath = chrome119.GetExecutablePath(),
                         }))
            {
                await using var page = await browser.NewPageAsync();
                await page.GoToAsync("https://www.whatismybrowser.com/");

                Console.WriteLine("Generating PDF");
                await page.PdfAsync(Path.Combine(Directory.GetCurrentDirectory(), "119.pdf"));

                Console.WriteLine("Export completed");
            }
        }
    }
}
```

**Comment by piskov** [comment]
@kblok I will try to reproduce, and get back to you.

In our case it’s a custom 30+ pages PDF report with headers and footers with the help of some logic that is like [puppeteer report](https://github.com/PejmanNik/puppeteer-report-csharp). It basically generates 3 pdf and merges them together.

We don’t actually use the lib but the [bundle.js](https://github.com/PejmanNik/puppeteer-report#faq) to slice the pages, generate three of them and then merge into one to have pages with custom headers and footers.

**Comment by piskov** [comment]
Just for sake of example of what kind of reports we’re generating: it’s heavy on svg, custom graphics etc. Generates around 4 seconds and 500 KB. With just setting chrome to 199 it goes to almost 12 seconds and 2 MB of size.

<img width="921" alt="Screenshot 2023-12-20 at 16 18 11" src="https://github.com/hardkoded/puppeteer-sharp/assets/4349021/fe2ffd27-75a8-4425-b9a0-3b44385ba960">

**Comment by kblok** [comment]
Can you reproduce that just by using the report. We could try reporting that in [chromium bugs](https://bugs.chromium.org/p/chromium/issues/list).
In the mean time, my exampled showed you how to use chrome 118.

**Comment by piskov** [comment]
Yeah we forced the 118, thank you: we actually implemented the “forcing” of chrome version in our config so that such a thing woudn’t happen :-)

That’s how I pinned the bug in the first place: I was fidgeting with versions of chrome and puppeteer to find the culprit.

**Comment by piskov** [comment]
@kblok why oh why have you bestowed such a mischief upon me :-)

I’ve talked about 119.0.**6045**.105 — the same you’ve used in your PR — yet in your sample you used 119.0.**5997**.0 Took me almost two hours to find out that’s the reason you got the same results.

tldr; your sample is fine. Change two things:
- use correct version `119.0.6045.105`
- navigate to some heavy website like `https://www.apple.com/ios/ios-17/`

What you will see is another very-very bad problem: file size and number of pages will be different, as is layout
<img width="1011" alt="Screenshot 2023-12-22 at 19 52 58" src="https://github.com/hardkoded/puppeteer-sharp/assets/4349021/f898843d-a53d-4b3f-bf63-90ca72123198">

… even with explicitly settting the options:
```
var pdfOptions = new PdfOptions
{
    PreferCSSPageSize = true,
    Format = PaperFormat.A4,
    PrintBackground = true,
    MarginOptions = new MarginOptions
    {
        Top = "0px", Left = "0px", Right = "0px", Bottom = "0px",   
    },
};
```

What it troubling: in our case (not the apple webpage render) we don’t see any layout or font size changes. The issue is very different with enourmous file size and time generation.

**Comment by piskov** [comment]
For us, issue is almost completely resolved with `122.0.6200.0`: now −4% file size and “only” 10% slower generation. Which is still a downgrade and the reason to stick with 118.

But the issue with apple website above is just the same even on 122, so the original proposal stands: #2354 should be reverted and fixed on 118 until chrome team figure this thing out.

**Comment by kblok** [comment]
I was trying to find something to report on chromium bugs. But I don't see anything that calls my attention here:

```
var chrome118Path = "/mac_arm-118.0.5993.70/chrome-mac-arm64/Google Chrome for Testing.app/Contents/MacOS/Google Chrome for Testing";
var chrome119Path = "/mac_arm-119.0.6045.105/chrome-mac-arm64/Google Chrome for Testing.app/Contents/MacOS/Google Chrome for Testing";
var chrome122Path = "/mac_arm-122.0.6200.0/chrome-mac-arm64/Google Chrome for Testing.app/Contents/MacOS/Google Chrome for Testing";

foreach (var browserInfo in new[]
     {
         new { Name = "chrome118", Path = chrome118Path },
         new { Name = "chrome119", Path = chrome119Path },
         new { Name = "chrome122", Path = chrome122Path },
     })
{
    Console.WriteLine($"Testing {browserInfo.Name}...");
    await using var browser = await Puppeteer.LaunchAsync(new LaunchOptions
    {
        ExecutablePath = browserInfo.Path,
    });

    var page = await browser.NewPageAsync();
    await page.GoToAsync("https://www.apple.com/ios/ios-17/");

    var stopwatch = new Stopwatch();
    stopwatch.Start();

    await page.PdfAsync($"{browserInfo.Name}.pdf", new PdfOptions()
    {
        Format = PaperFormat.A4,
        PrintBackground = true,
        MarginOptions = new MarginOptions
        {
            Top = "0px",
            Bottom = "0px",
            Left = "0px",
            Right = "0px",
        },
    });
    Console.WriteLine($"PDF generation took {stopwatch.ElapsedMilliseconds}ms");
}
```

```
Testing chrome118...
PDF generation took 263ms
Testing chrome119...
PDF generation took 153ms
Testing chrome122...
PDF generation took 161ms
```

![image](https://github.com/hardkoded/puppeteer-sharp/assets/2198466/2e80df96-2069-4fc3-ac76-3acaa20d9101)

> But the issue with apple website above is just the same even on 122, so the original proposal stands: https://github.com/hardkoded/puppeteer-sharp/pull/2354 should be reverted and fixed on 118 until chrome team figure this thing out.

I don't think we have to rollback back, having a way to workaround it.

**Comment by piskov** [comment]
Tested your code on my macbook air m1 with sonoma 14.3 public beta — see the difference in file sizes starting with 119:

<img width="280" alt="Screenshot 2023-12-23 at 02 00 53" src="https://github.com/hardkoded/puppeteer-sharp/assets/4349021/7050c5af-f737-4f32-8811-885a50f3bc3f">

118 has 24 pages
119 and later has 28

The only difference with yours are filepaths for chrome (I downloaded them via code like `browserFetcher.DownloadAsync("118.0.5993.70")`)

```
var chrome118Path = "/Users/piskov/Downloads/ChromePdfTest/bin/Release/net8.0/Chrome/MacOSArm64-118.0.5993.70/chrome-mac-arm64/Google Chrome for Testing.app/Contents/MacOS/Google Chrome for Testing";
var chrome119Path = "/Users/piskov/Downloads/ChromePdfTest/bin/Release/net8.0/Chrome/MacOSArm64-119.0.6045.105/chrome-mac-arm64/Google Chrome for Testing.app/Contents/MacOS/Google Chrome for Testing";
var chrome122Path = "/Users/piskov/Downloads/ChromePdfTest/bin/Release/net8.0/Chrome/MacOSArm64-122.0.6200.0/chrome-mac-arm64/Google Chrome for Testing.app/Contents/MacOS/Google Chrome for Testing";
```

Here are the PDFs:
[chrome118.pdf](https://github.com/hardkoded/puppeteer-sharp/files/13757234/chrome118.pdf)
[chrome119.pdf](https://github.com/hardkoded/puppeteer-sharp/files/13757235/chrome119.pdf)
[chrome122.pdf](https://github.com/hardkoded/puppeteer-sharp/files/13757236/chrome122.pdf)


How come yours are the same?

**Comment by piskov** [comment]
Anyhoo, my issue is resolved with chrome 122, and you’ll update to it at some point.

While I am troubled with the difference of the PDF rendering of the Apple website (different margins, font sizes, number of pages, etc.), it fortunately doesn’t affect the docs we are creating.

Thank you for your time and effort — I really appreciate it! And If you would like me to check some things more (like I’ve demonstrated above, I can replicate the same differences of rendering both on x64 win 11 and m1 mac), please jot me a line.

**Comment by kblok** [comment]
Awesome! I'll create a new version when we have a new Chrome release and I'll let you know.
